File: leetcode/0300-longest-increasing-subsequence/solution.py
import bisect
from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        tails = []  # tails[i] = smallest possible tail value of an increasing subsequence of length i+1

        for n in nums:
            pos = bisect.bisect_left(tails, n)
            if pos == len(tails):
                tails.append(n)
            else:
                tails[pos] = n

        return len(tails)

File: leetcode/0300-longest-increasing-subsequence/Solution.java
import java.util.Arrays;

class Solution {
    public int lengthOfLIS(int[] nums) {
        int[] tails = new int[nums.length];
        int size = 0;

        for (int n : nums) {
            int lo = 0, hi = size;
            while (lo < hi) {
                int mid = (lo + hi) / 2;
                if (tails[mid] < n) {
                    lo = mid + 1;
                } else {
                    hi = mid;
                }
            }
            tails[lo] = n;
            if (lo == size) {
                size++;
            }
        }

        return size;
    }
}
